fix(datetimepickerv2): add single select and replace range select with new timepickerdropdown

jessieyan opened the pull request:
Closes #3503 

**Summary**

- add single select
- add new prop useNewTimeSpinner, if enabled, absolute will use `<TimePickerDropdown> `as the time picker. Single select only works when useNewTimeSpinner is enabled. 

**Change List (commits, features, bugs, etc)**

` <DateTimePickerV2WithTimeSpinner>` is the component that use` <TimePickerDropdown>`
`<DateTimePickerV2WithoutTimeSpinner>` is the previous `<DateTimePickerV2> `
`<DateTimePickerV2>` is the wrapper. If `useNewTimeSpinner `enabled,  it renders `<DateTimePickerV2WithTimeSpinner>  `otherwise renders `<DateTimePickerV2WithoutTimeSpinner>`

**Acceptance Test (how to verify the PR)**

Single select
![image](https://user-images.githubusercontent.com/24279794/186203230-7d6b3f1d-e72a-4a5d-8e58-d70584f9627c.png)

Range select
![image](https://user-images.githubusercontent.com/24279794/186203527-93b92322-4b63-4449-89c1-01874d005d59.png)


**Regression Test (how to make sure this PR doesn't break old functionality)**

- tests here

<!-- For reviewers: do not remove -->

**Things to look for during review**

- [ ] Make sure all references to `iot` or `bx` class prefix is using the prefix variable
- [ ] (React) All major areas have a `data-testid` attribute. New test ids should have test written to ensure they are not changed or removed.
- [ ] UI should be checked in RTL mode to ensure the proper handling of layout and text.
- [ ] All strings should be translatable.
- [ ] The code should pass a11y scans (The storybook a11y knob should show no violations). New components should have a11y test written.
- [ ] Unit test should be written and should have a coverage of 90% or higher in all areas.
- [ ] All components should be passing visual regression test. For new styles or components either a visual regression test should be written for all permutations or the base image updated.
- [ ] Changes or new components should either write new or update existing documentation.
- [ ] PR should link and close out an existing issue

davidicus commented on packages/react/src/components/DateTimePicker/DateTimePickerV2.test.e2e.jsx:
```diff
@@ -107,24 +136,25 @@ describe('DateTimePickerV2', () => {
     cy.findByText(i18n.applyBtnLabel).should('be.disabled');
-
-    // set time to 11:00
-    cy.findByLabelText(i18n.endTimeLabel).type('{backspace}{backspace}00');
-    cy.findByText(i18n.applyBtnLabel).should('not.be.disabled');
```
is there a reason we took this out?

jessieyan replied:
> is there a reason we took this out?

This is the behavior in <TimePickerSpinner> that if put 00 in the input, it is not a invalid value. Now we use <TimePickerDropDown> it would cause invalid date error.

davidicus commented on packages/react/src/components/TimePicker/TimePickerDropdown.jsx:
```diff
@@ -297,7 +302,7 @@ const TimePickerDropdown = ({
             <TextInput
               onChange={handleOnChange}
               data-testid={`${testId}-input`}
-              id={id}
```
changing id would be considered a breaking change.

davidicus commented on packages/react/src/components/TimePicker/TimePickerDropdown.jsx:
```diff
@@ -364,7 +369,7 @@ const TimePickerDropdown = ({
               data-testid={`${testId}-input-1`}
               onFocus={() => setFocusedInput(0)}
-              id={`${id}-1`}
+              id={`${id}-start-time`}
```
same comment about the id

davidicus commented on packages/react/src/components/TimePicker/TimePickerDropdown.jsx:
```diff
@@ -403,7 +408,7 @@ const TimePickerDropdown = ({
             <TextInput
               data-testid={`${testId}-input-2`}
-              id={`${id}-2`}
+              id={`${id}-end-time`}
```
breaking change

jessieyan replied:
I will change the ids back.

JordanWSmith15 reviewed (changes requested):
Tested on Mac Firefox
1. When I focus on the time picker and enter a number, focus is lost
2. I see "Invalid date" fairly often. 
a. It should give a standard invalid warning (red text below the input) if it's invalid, but still show what I selected
b. It appears seemingly for no reason (Hitting Apply)
![Screen Shot 2022-08-23 at 3 47 34 PM](https://user-images.githubusercontent.com/21066152/186252608-4334bd2f-2fa7-43b4-8f9b-2b89cb96f8eb.png)

davidicus commented on packages/react/src/components/TimePicker/_time-picker-dropdown.scss:
```diff
@@ -66,14 +66,17 @@
 
     .#{$iot-prefix}--time-picker__wrapper-sm & {
       bottom: 1rem;
+      padding-right: 0rem;
```
nit: zero values should have no units

davidicus commented:
I try to open the time picker spinner dropdown and it does not open on [this](https://deploy-preview-3523--carbon-addons-iot-react.netlify.app/?path=/story/2-watson-iot-experimental-%E2%98%A2%EF%B8%8F-datetimepickerv2--default) story.

davidicus reviewed (changes requested):
Some issues to address.

jessieyan commented on packages/react/src/components/DateTimePicker/DateTimePickerV2.test.jsx:
```diff
@@ -825,46 +932,85 @@ describe('DateTimePickerV2', () => {
     expect(screen.getByText(i18n.applyBtnLabel)).toBeEnabled();
 
     // 2020-04-01 13:34 to 2020-04-06 11:49
-    userEvent.type(startTime, '{backspace}{backspace}{backspace}{backspace}{backspace}13:34');
-    expect(startTime).toHaveValue('13:34');
-    expect(endTime).toHaveValue('11:49');
-    expect(timeRange).toHaveTextContent('2020-04-01 13:34 to 2020-04-06 11:49');
```
> I see many of these old test deleted. Should we not keep these old test and add the new ones to ensure we are not breaking past functionality?

userEvent.type() does not work for some reason. I have to use fireEvent.chanage() to change a value. I use click on the spinners instead for the tests.

jessieyan commented:
> 2\. below



> Tested on Mac Firefox
> 
> 1. When I focus on the time picker and enter a number, focus is lost
> 2. I see "Invalid date" fairly often.
>    a. It should give a standard invalid warning (red text below the input) if it's invalid, but still show what I selected
>    b. It appears seemingly for no reason (Hitting Apply)
>    ![Screen Shot 2022-08-23 at 3 47 34 PM](https://user-images.githubusercontent.com/21066152/186252608-4334bd2f-2fa7-43b4-8f9b-2b89cb96f8eb.png)

For 2, I will update the story to expose the warning text. 
![image](https://user-images.githubusercontent.com/24279794/186254468-3186542a-03d3-4f30-839a-554c64df19cc.png)

jessieyan commented:
> 



> Tested on Mac Firefox
> 
> 1. When I focus on the time picker and enter a number, focus is lost
> 2. I see "Invalid date" fairly often.
>    a. It should give a standard invalid warning (red text below the input) if it's invalid, but still show what I selected
>    b. It appears seemingly for no reason (Hitting Apply)
>    ![Screen Shot 2022-08-23 at 3 47 34 PM](https://user-images.githubusercontent.com/21066152/186252608-4334bd2f-2fa7-43b4-8f9b-2b89cb96f8eb.png)

For 1, it's some logic from the existing <DateTimePickerV2> . Will take another look.

JordanWSmith15 commented:
> > 
> 
> > Tested on Mac Firefox
> > 
> > 1. When I focus on the time picker and enter a number, focus is lost
> > 2. I see "Invalid date" fairly often.
> >    a. It should give a standard invalid warning (red text below the input) if it's invalid, but still show what I selected
> >    b. It appears seemingly for no reason (Hitting Apply)
> >    ![Screen Shot 2022-08-23 at 3 47 34 PM](https://user-images.githubusercontent.com/21066152/186252608-4334bd2f-2fa7-43b4-8f9b-2b89cb96f8eb.png)
> 
> For 1, it's some logic from the existing . Will take another look.

This is what I'd expect for 2A:
![Screen Shot 2022-08-23 at 4 09 58 PM](https://user-images.githubusercontent.com/21066152/186256249-13c74fea-f232-4891-9070-9a369750c566.png)

jessieyan commented:
> I try to open the time picker spinner dropdown and it does not open on [this](https://deploy-preview-3523--carbon-addons-iot-react.netlify.app/?path=/story/2-watson-iot-experimental-%E2%98%A2%EF%B8%8F-datetimepickerv2--default) story.

Put zIndex setting is the story. Now it should work.

jessieyan commented:
> > > 
> > 
> > 
> > > Tested on Mac Firefox
> > > 
> > > 1. When I focus on the time picker and enter a number, focus is lost
> > > 2. I see "Invalid date" fairly often.
> > >    a. It should give a standard invalid warning (red text below the input) if it's invalid, but still show what I selected
> > >    b. It appears seemingly for no reason (Hitting Apply)
> > >    ![Screen Shot 2022-08-23 at 3 47 34 PM](https://user-images.githubusercontent.com/21066152/186252608-4334bd2f-2fa7-43b4-8f9b-2b89cb96f8eb.png)
> > 
> > 
> > For 1, it's some logic from the existing . Will take another look.
> 
> This is what I'd expect for 2A: ![Screen Shot 2022-08-23 at 4 09 58 PM](https://user-images.githubusercontent.com/21066152/186256249-13c74fea-f232-4891-9070-9a369750c566.png)

1. The focus should be fixed
2. The invalid text should now show under the start/end time input.

jessieyan commented:
The enter key problem should be fixed.

jessieyan commented:
I'm fixing the firefox browser issue that Jordan found. I need to format the hh:mm A date without usind dayjs.

jessieyan commented:
Updated some formats to make sure it works with Firefox and other browsers.

JordanWSmith15 reviewed (changes requested):
Two more things and then I think we're close:
1. I need to click the "Apply" primary button twice before the dialog closes
2. Null values don't seem to be demonstrated.
A. What does a null value look like, exactly? I would expect that placeholder grey text would show the browser's locale format:
![Screen Shot 2022-08-29 at 8 15 38 AM](https://user-images.githubusercontent.com/21066152/187199141-4e6bb6b0-b736-4530-ba5b-1c8f272742dd.png)
B. There is no way to clear the field back to null. I actually don't even see what the field looks like when null. 
<img width="334" alt="Screen Shot 2022-08-29 at 8 03 53 AM" src="https://user-images.githubusercontent.com/21066152/187197022-089746f3-19aa-48e7-a6d2-60ca3e93b5b1.png">

JordanWSmith15 commented:
@jessieyan 
1. Maybe it's specific to Mac? Do you have a PC? 
2. I'm talking specifically about being able to clear the "Single select" version's value. Graphite will not be using the Relative version for this release, so I am very focused on this component: 
https://deploy-preview-3523--carbon-addons-iot-react.netlify.app/?path=/story/2-watson-iot-experimental-%E2%98%A2%EF%B8%8F-datetimepickerv2--single-select

jessieyan commented:
Added clear button to reset date and time value in single select.

JordanWSmith15 reviewed (approved):
🥳 🎉

davidicus commented:
@jessieyan A few items:

- It looks like you took away the `TimePickerSpinner` option. The new input options were supposed to be put behind a prop. This component should work exactly as it did before unless this prop is provided. 
- Test for the old 24hour format should remain to make sure the old functionality still works. We should write additional test to test this new TimePickerDropdown functionality.
- When you are going from a [relative option with an existing value](https://deploy-preview-3523--carbon-addons-iot-react.netlify.app/?path=/story/2-watson-iot-experimental-%E2%98%A2%EF%B8%8F-datetimepickerv2--selected-relative) to the absolute value the existing value is being turned to "invalid date to invalid date". This remains even when you go back to the relative option.  
- When you go to use the absolute value option on any [story](https://deploy-preview-3523--carbon-addons-iot-react.netlify.app/?path=/story/2-watson-iot-experimental-%E2%98%A2%EF%B8%8F-datetimepickerv2--default) the input states "invalid date to invalid date" and the inputs are in the error state. This should not occur unless the user inputs an invalid date. 
- I am not seeing a story knob for switching between single and range or to use the old `TImePickerSpinner` vs. the new `TimePickerDropdown`

davidicus commented on packages/react/src/components/DateTimePicker/dateTimePickerUtils.js:
```diff
@@ -7,6 +7,30 @@ import dayjs from '../../utils/dayjs';
 
 const { iotPrefix } = settings;
 
+const customParseFormat = require('dayjs/plugin/customParseFormat');
```
Should this not be imported

davidicus commented on packages/react/src/components/TimePicker/TimePickerDropdown.jsx:
```diff
@@ -215,13 +245,8 @@ const TimePickerDropdown = ({
-      // close dropdown and validate
+      // close dropdown
       setOpenState(false);
-      setInvalidState(!validate(inputRef.current.value));
```
The removal of this code now allows for values other than numbers to be entered into the input.

davidicus replied:
Go to [this](https://next.carbon-addons-iot-react.com/?path=/story/1-watson-iot-timepickerdropdown--default) story and try typing in letters. With your changes the [same story](https://deploy-preview-3523--carbon-addons-iot-react.netlify.app/?path=/story/1-watson-iot-timepickerdropdown--default) now allows for these inputs to go through.

davidicus commented on packages/react/src/components/TimePicker/TimePickerDropdown.test.jsx:
```diff
@@ -197,7 +197,6 @@ describe('TimePickerDropdown', () => {
     expect(qbt('time-picker-test-spinner')).toBeInTheDocument();
     userEvent.click(document.body);
     await waitFor(() => expect(qbt('time-picker-test-spinner')).not.toBeInTheDocument());
-    expect(screen.queryByText(/The time entered is invalid/)).toBeTruthy();
```
This should not be removed. If all these steps in this test occur, this assertion should be truthy.

davidicus reviewed (changes requested):
There are a few items that still need to be addressed

jessieyan commented:
@EdmondCShaffer @herleraja  Hey guys, we want to the replace `<TimePickerSpinner> `with `< TimePickerDropDown>` in `<DateTimePickerV2>`. We need your feedback if you would like the keep` <TimePickerSpinner>`  or using the new ` < TimePickerDropDown>`.

The input and output of <DateTimePickerV2> will stay the same. Main differences are `<TimePickerDropDown> `uses a list spinner for hours, minutes and AM/PM. Also `<TimePickerDropDown>` uses a 12 hr format for displaying the time. 

Please let us know your preference and concerns.

herleraja commented:
> DateTimePickerV2

He @jessieyan I think its fine. Only to note that we use the 24-hour format in the dashboard card DateTimePickerV2. I think by default its 24 hours then we don't have much concern here.

Btw what is the difference between DateTimePickerV2 and DateTimePicker?

jessieyan commented:
@davidicus I renamed old `<DateTimePickerV2>` to `< DateTimePickerV2WithoutTimeSpinner>` and use `<DateTimePickerV2>` to wrap  `< DateTimePickerV2WithoutTimeSpinner>`  and  `< DateTimePickerV2WithTimeSpinner>`. I saw the PublicApi show some red lines. Should I define all props in the wrapper `<DateTimePickerV2> `as well?

jessieyan commented:
> @davidicus I renamed old `<DateTimePickerV2>` to `< DateTimePickerV2WithoutTimeSpinner>` and use `<DateTimePickerV2>` to wrap `< DateTimePickerV2WithoutTimeSpinner>` and `< DateTimePickerV2WithTimeSpinner>`. I saw the PublicApi show some red lines. Should I define all props in the wrapper `<DateTimePickerV2> `as well?

Fixed the the `propTypes` and `defaultProps` in `<DateTimePickerV2>`.

jessieyan commented:
> @jessieyan A few items:
> 
> * It looks like you took away the `TimePickerSpinner` option. The new input options were supposed to be put behind a prop. This component should work exactly as it did before unless this prop is provided.
> * Test for the old 24hour format should remain to make sure the old functionality still works. We should write additional test to test this new TimePickerDropdown functionality.
> * When you are going from a [relative option with an existing value](https://deploy-preview-3523--carbon-addons-iot-react.netlify.app/?path=/story/2-watson-iot-experimental-%E2%98%A2%EF%B8%8F-datetimepickerv2--selected-relative) to the absolute value the existing value is being turned to "invalid date to invalid date". This remains even when you go back to the relative option.
> * When you go to use the absolute value option on any [story](https://deploy-preview-3523--carbon-addons-iot-react.netlify.app/?path=/story/2-watson-iot-experimental-%E2%98%A2%EF%B8%8F-datetimepickerv2--default) the input states "invalid date to invalid date" and the inputs are in the error state. This should not occur unless the user inputs an invalid date.
> * I am not seeing a story knob for switching between single and range or to use the old `TImePickerSpinner` vs. the new `TimePickerDropdown`
> * When I click enter on the TimePickerDropdown component, inside DateTimePicker, I get an error thrown
>   ![image](https://user-images.githubusercontent.com/1590966/187998680-15061d3f-927c-4747-9890-9a8d7c708bfd.png)

@davidicus 
This should be addressed by adding a new prop `useNewTimeSpinner` and the wrapper component <DateTimePickerV2> which determines whether to render <DateTimePickerV2WithTimeSpinner> or <DateTimePickerV2WithoutTimeSpinner>. Added knob of `useNewTimeSpinner` in `Selected absolute` story and add a separate story Range select with new time spinner. 

![image](https://user-images.githubusercontent.com/24279794/188724337-d7e8b3fe-9a60-4682-85d8-834fc284d1d3.png)

davidicus commented on packages/react/src/components/DateTimePicker/DateTimePickerV2.test.jsx:
```diff
@@ -528,47 +1047,133 @@ describe('DateTimePickerV2', () => {
+    expect(screen.getByText('minutes')).toBeInTheDocument();
   });
 
-  it('should not render a time picker if hasTimeInput:false', () => {
```
There still seems to be changed test here. These should stay to ensure the old functionality is being preserved, no?

jessieyan replied:
> There still seems to be changed test here. These should stay to ensure the old functionality is being preserved, no?

The test was moved to line 1087.

davidicus commented on packages/react/src/components/TimePicker/TimePickerDropdown.jsx:
```diff
@@ -183,6 +186,12 @@ const TimePickerDropdown = ({
     }
   };
 
+  const handleOnKeyUp = (e) => {
+    if (e.key === 'Enter') {
+      setOpenState(true);
+    }
+  };
+
```
This should not be applied to this component as it breaks the ability for [implicit submission](https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#implicit-submission). The icon is what allows for the dropdown functionality on enter and is separate tab stop than the text input.

jessieyan replied:
> This should not be applied to this component as it breaks the ability for [implicit submission](https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#implicit-submission). The icon is what allows for the dropdown functionality on enter and is separate tab stop than the text input.

There is a bug in <DateTimePickerV2>. The keyup event can detect the enter key but the keydown event did not detect.

davidicus commented on packages/react/src/components/TimePicker/_time-picker-dropdown.scss:
```diff
@@ -66,14 +66,17 @@
 
     .#{$iot-prefix}--time-picker__wrapper-sm & {
       bottom: 1rem;
+      padding-right: 0;
```
I believe that this element already has no padding applied. So these padding declarations are unnecessary.

jessieyan replied:
Moved the padding-right out of <TimePickerDropdown>.

davidicus commented:
In [this](https://deploy-preview-3523--carbon-addons-iot-react.netlify.app/?path=/story/2-watson-iot-experimental-%E2%98%A2%EF%B8%8F-datetimepickerv2--single-select) story you can see the tooltip value is showing the mask value instead of the value of the input as it does in our current version. 
Next             |  This PR
:-------------------------:|:-------------------------:
![](https://user-images.githubusercontent.com/1590966/189439349-da81602f-9b77-4127-aeb3-91f26ba0b199.png)  |  ![](https://user-images.githubusercontent.com/1590966/189439211-ca53976b-3fb5-466f-940b-dc78e7b91ef3.png)

davidicus commented:
This may not something that we need to fix now but in the input when you click a on a date from an empty state you get a time as seen in this image. Should this value be added as the value for the TimePickerDropdown? ![image](https://user-images.githubusercontent.com/1590966/189439743-0a87d35f-4ac2-4ded-a0f4-c7fd1c97a884.png)

davidicus commented:
In [this](https://deploy-preview-3523--carbon-addons-iot-react.netlify.app/?path=/story/2-watson-iot-experimental-%E2%98%A2%EF%B8%8F-datetimepickerv2--selected-absolute) story when you go to a preset value then chose a relative value then go to the absolute value calendar, there is a discrepancy in what is in the input field and what is shown in the `TImePickerDropdown`
![image](https://user-images.githubusercontent.com/1590966/189441471-2c723409-9d5e-483d-a5da-6fe6ff9e6f4f.png)

jessieyan commented:
> this



> In [this](https://deploy-preview-3523--carbon-addons-iot-react.netlify.app/?path=/story/2-watson-iot-experimental-%E2%98%A2%EF%B8%8F-datetimepickerv2--single-select) story you can see the tooltip value is showing the mask value instead of the value of the input as it does in our current version.
> 
> Next	This PR
> ![](https://user-images.githubusercontent.com/1590966/189439349-da81602f-9b77-4127-aeb3-91f26ba0b199.png)	![](https://user-images.githubusercontent.com/1590966/189439211-ca53976b-3fb5-466f-940b-dc78e7b91ef3.png)

This is the original design I think. The design before is if the whole value could not be displayed, then show it in tooltip. If the whole value could be shown, then show mask value in the tooltip.

davidicus commented:
@jessieyan there may be something in the code that does not allow for this check to be dynamic but when I make the input smaller I am not getting the value in the tool tip. Can we check if this is the case and what we can do about it. I think this would be the last thing and then we can get this merged in. Thanks for working through this. I know its a complicated component. 
![image](https://user-images.githubusercontent.com/1590966/189687189-da441671-de81-420f-a411-d3fae978d1dc.png)

jessieyan commented:
> In [this](https://deploy-preview-3523--carbon-addons-iot-react.netlify.app/?path=/story/2-watson-iot-experimental-%E2%98%A2%EF%B8%8F-datetimepickerv2--selected-absolute) story when you go to a preset value then chose a relative value then go to the absolute value calendar, there is a discrepancy in what is in the input field and what is shown in the `TImePickerDropdown` ![image](https://user-images.githubusercontent.com/1590966/189441471-2c723409-9d5e-483d-a5da-6fe6ff9e6f4f.png)
> 
> @davidicus Yeah, there is a bug here. This bug is not new though. When change time values in relative, they are not picked up by absolute. I will see if i can fix it. Otherwise, I think we could open an issue to track it.

Hi @davidicus , I opened issue [3547](https://github.com/carbon-design-system/carbon-addons-iot-react/issues/3547) to track the issue that values are not updated when switching between Relative and Absolute.
